rasa_core/server.py

```diff
@@ -288,6 +290,50 @@ def parse(self, request, sender_id):
                          "parse: {}".format(e), exc_info=1)
             return json.dumps({"error": "{}".format(e)})
 
+    @app.route("/conversations/<sender_id>/respond",
+               methods=['GET', 'POST', 'OPTIONS'])
+    @check_cors
+    @requires_auth
+    @ensure_loaded_agent
+    def respond(self, request, sender_id):
+        request.setHeader('Content-Type', 'application/json')
+        if request.method.decode('utf-8', 'strict') == 'GET':
+            request_params = {
+                key.decode('utf-8', 'strict'): value[0].decode('utf-8',
+                                                               'strict')
+                for key, value in request.args.items()}
+        else:
+            content = request.content.read()
+            try:
+                request_params = json.loads(content.decode('utf-8', 'strict'))
+            except ValueError as e:
+                logger.error("Failed to decode json during parse request. "
+                             "Error: {}. Request content: "
+                             "'{}'".format(e, content))
+                raise
+
+        if 'query' in request_params:
+            message = request_params.pop('query')
+        elif 'q' in request_params:
+            message = request_params.pop('q')
+        else:
+            request.setResponseCode(400)
+            return json.dumps({"error": "Invalid parse parameter specified"})
+
+        try:
+            out = CollectingOutputChannel()
+            processor = self.agent._create_processor()
```

Comment:
let's directly use `self.agent.handle_message` instead to avoid this private call.

Reply:
that doesn't work because `handle_message` takes only text as input, we need a `UserMessage` so that we can specify the output channel & sender id. Happy to modify the handle_message method to take either though

Reply:
`handle_message` should already accept the output channel:

```
    def handle_message(
            self,
            text_message,  # type: Text
            message_preprocessor=None,  # type: Optional[Callable[[Text], Text]]
            output_channel=None,  # type: Optional[OutputChannel]
            sender_id=UserMessage.DEFAULT_SENDER_ID  # type: Optional[Text]
    ):
        # type: (...) -> Optional[List[Text]]
        """Handle a single message.
```

Reply:
oh nice!